src/vcpkg/base/downloads.cpp

```diff
@@ -110,9 +214,30 @@ namespace vcpkg
                 ret.m_hRequest = WinHttpHandle{h};
             }
 
+            if (maybe_proxy_creds)
+            {
+                const auto& proxy_creds = maybe_proxy_creds.value_or_exit(VCPKG_LINE_INFO);
+                ret.m_proxy_credentials = WinHttpContext(proxy_creds);
+
+                WinHttpSetCredentials(ret.m_hRequest.h,
+                                      WINHTTP_AUTH_TARGET_PROXY,
+                                      WINHTTP_AUTH_SCHEME_BASIC,
+                                      proxy_creds.username.data(),
+                                      proxy_creds.password.data(),
+                                      NULL);
+
+                WinHttpSetStatusCallback(
+                    ret.m_hRequest.h, &set_proxy_credentials_on_redirect, WINHTTP_CALLBACK_FLAG_REDIRECT, NULL);
+            }
+
             // Send a request.
-            auto bResults = WinHttpSendRequest(
-                ret.m_hRequest.h, WINHTTP_NO_ADDITIONAL_HEADERS, 0, WINHTTP_NO_REQUEST_DATA, 0, 0, 0);
+            auto bResults = WinHttpSendRequest(ret.m_hRequest.h,
+                                               WINHTTP_NO_ADDITIONAL_HEADERS,
+                                               0,
+                                               WINHTTP_NO_REQUEST_DATA,
+                                               0,
+                                               0,
+                                               ret.m_proxy_credentials.as_dword_ptr());
```

Comment:
Given that this data is tied to the lifetime of the request, is there any reason we don't just use an ordinary member variable and pass `reinterpret_cast<DWORD_PTR>(this)` for that?

Then `set_proxy_credentials_on_redirect` does `reinterpret_cast<WinHttpRequest*>(pContext)->any data you want`.